crates/rrg/src/action/yara_scan_memory.rs

```diff
@@ -0,0 +1,564 @@
+            Ok(regions) => regions,
+            Err(cause) => {
+                session.reply(Err(ErrorItem {
+                    pid,
+                    error: Error::OpenProcessMemory(cause),
+                }))?;
+                continue;
+            }
+        };
+        let mut memory = match ReadableProcessMemory::open(pid) {
+            Ok(memory) => memory,
+            Err(cause) => {
+                session.reply(Err(ErrorItem {
+                    pid,
+                    error: Error::OpenProcessMemory(cause),
+                }))?;
+                continue;
+            }
+        };
+
+        // ParseRegionErrors are internal errors, so wrap them in `Error::action`
+        // and bail early if any error is encountered.
+        let regions: Vec<MappedRegion> = regions
+            .map(|reg| reg.map_err(crate::session::Error::action))
+            .collect::<Result<_, _>>()?;
+
+        let mut scanner = Scanner::new(&rules);
+        if let Some(timeout) = args.scan_timeout {
+            scanner.set_timeout(timeout);
+        }
+
+        if let Err(error) = regions
+            .into_iter()
+            .filter(|reg| args.should_dump(reg))
+            .try_for_each(|region| {
+                scan_region(
+                    &region,
+                    &mut scanner,
+                    &mut memory,
+                    args.chunk_size,
+                    args.chunk_overlap,
+                )
+            })
+        {
+            session.reply(Err(ErrorItem { pid, error }))?;
+            continue;
+        }
+
+        match scanner.finish() {
+            Ok(results) => {
+                let matching_rules = results
+                    .matching_rules()
+                    .map(|rule| Rule::blobify(rule, session))
+                    .collect::<crate::session::Result<Vec<_>>>()?;
+                session.reply(Ok(OkItem {
+                    pid,
+                    matching_rules,
+                }))?;
+            }
+            Err(error) => {
+                session.reply(Err(ErrorItem {
+                    pid,
+                    error: Error::Scan(error),
+                }))?;
+            }
+        }
+    }
+    Ok(())
+}
+
+#[cfg(not(any(target_os = "linux", target_os = "windows")))]
+pub fn handle<S>(_session: &mut S, _args: Args) -> crate::session::Result<()>
+where
+    S: crate::session::Session,
+{
+    use std::io::{Error, ErrorKind};
+    Err(crate::session::Error::action(Error::from(
+        ErrorKind::Unsupported,
+    )))
+}
+
+#[cfg(all(test, any(target_os = "linux", target_os = "windows")))]
+mod tests {
+    use super::*;
+    use crate::action::dump_process_memory::FakeProcessMemory;
+
+    const EXAMPLE_RULE_SRC: &str = r#"
+    rule ExampleRule {
+        strings:
+            $text = "text here"
+            $hex = { E2 34 A1 C8 23 FB }
+            $regex = /some regular expression: \w+/
+        condition:
+            $text or $hex or $regex
+    }
+    "#;
```

Comment:
I would prefer not to have such constants, forcing the reader to jump around when going over the test.